File: cs12b/lab5/Dictionary.c
//Shayan Salsabilian
//ssalsabi
//2/19/19
//12B
//Implementation file for Dictionary.h
//Dictionary.c
#include<stdio.h>
#include<stdlib.h>
#include<string.h>
#include<assert.h>
#include"Dictionary.h"


// private types --------------------------------------------------------------

// NodeObj
typedef struct NodeObj
{
  char *key;			//creates a key
  char *value;			//creates a value
  struct NodeObj *next;		//includes a next Obj in the linked list
} NodeObj;

// Node
typedef NodeObj *Node;		//makes the NodeObj a node

// newNode()
// constructor of the Node type
Node
newNode (char *key, char *value)
{				//makes a newNode with a specific key and value needed
  Node N = malloc (sizeof (NodeObj));	//allocate an amount of memory to hold the NodeObj
  assert (N != NULL);		//if we have a null malloc return null
  N->key = key;			//place the key in the key area of the node object
  N->value = value;		//place the value in the value area of the node object
  N->next = NULL;		//make the next item the null
  return (N);			//return the node
}

// freeNode()
// destructor for the Node type
void
freeNode (Node * pN)
{
  if (pN != NULL && *pN != NULL)
    {				//make sure the node isn't already empty 
      free (*pN);		//if its not free the node
      *pN = NULL;		//then set its pointer equal to null so we cant find it
    }
}

// DictionaryObj
typedef struct DictionaryObj
{				//create a DictionaryObj
  Node head;			//make the head node that will point to the front of the list
  int numItems;			//keep track of the number of items
} DictionaryObj;


// public functions -----------------------------------------------------------

// newDictionary()
// constructor for the Dictionary type
// parameters: none
// post: Dictionary S
Dictionary
newDictionary (void)
{				//create a new dictionary
  Dictionary S = malloc (sizeof (DictionaryObj));	//make a Dictionary that allocates memory for the size of the DictionaryObj
  assert (S != NULL);		//check for malloc failures
  S->head = NULL;		//set the head to null
  S->numItems = 0;		//set the numItems to 0
  return S;			//return the dictionary
}

// freeDictionary()
// destructor for the Dictionary type
// parameters: Dictionary pS
// post: none
void
freeDictionary (Dictionary * pS)
{				//free the entire dictionary
  if (pS != NULL && *pS != NULL)
    {				//check to make sure its not a null dictionary
      if (!isEmpty (*pS))	//if the dictionary is not empty then empty it
	makeEmpty (*pS);
      free (*pS);		//free the dictionary
      *pS = NULL;		//set the pointer to null
    }
}

// isEmpty()
// returns 1 (true) if S is empty, 0 (false) otherwise
// parameters: none
// post: 1 if numItems is 0, and 0 if numItems is 1
int
isEmpty (Dictionary S)
{
  if (S == NULL)
    {				//if null print a null dictionary reference error
      fprintf (stderr,
	       "Dictionary Error: isEmpty() called  on NULL Dictionary reference\n");
      exit (EXIT_FAILURE);
    }
  return (S->numItems == 0);	//otherwise return one if the size is 0 and 0 otherwise
}

// size()
//returns number of items
//parameters: none
//post: numItems
int
size (Dictionary D)
{
  if (D == NULL)
    {
      fprintf
	(stderr,
	 "Dictionary Error: size() called on NULL Dictionary reference\n");
      exit (EXIT_FAILURE);
    }
  return D->numItems;		//just checks the number of items and returns it
}

//lookup()
//returns the value v such that (k,v) is in D, or returns null if not
//parameters:Dictionary D, char* k
//post: value contained char*
char *
lookup (Dictionary D, char *k)
{				//lookup the value needed for the key
  if (D == NULL)
    {				//if the dictionary pointer is null return a NULL dictionary reference error
      fprintf (stderr,
	       "Dictionary Error: lookup() called on NULL Dictionary reference\n");
      exit (EXIT_FAILURE);
    }				//otherwise
  Node N = D->head;		//save the start of the dictionary into Node N
  while (N != NULL)
    {				//then go until the end of the dictionary
      if ((strcmp (N->key, k)) == 0)
	{			//then string compare till we find the identical key
	  return N->value;	//then return the value when the key is identical
	}
      N = N->next;		//otherwise go to the next node in the dictionary
    }
  return NULL;			//if we can't find it return a null
}

//find() (private function)
//parameters:Dictionary D ,number of items in Dictionary
//post: node at the number of items
//finds the last node on the list
Node
find (Dictionary D, int numItems)
{
  if (D == NULL)
    {				//if the dictionary pointer is null return a NULL dictionary reference error
      fprintf (stderr,
	       "Dictionary Error: find() called on NULL Dictionary reference\n");
      exit (EXIT_FAILURE);
    }
  int i = 1;			//start the counter at 1
  Node N = D->head;		//start at the head of the dictionary
  while (i < numItems)
    {				//go till we reach the number of items specified
      N = N->next;
      i++;
    }
  return N;			//once there return the node your at
}

//count() (private function)
//parameters: Dictionary D, key
//post: index
//find the index where the  key value is held
int
findIndex (Dictionary D, char *key)
{
  int count = 0;		//start the count
  Node N = D->head;		//put the head of the function in the Node
  while (count <= D->numItems)
    {				//increment through the Dictionary 
      if ((strcmp (key, N->key)) == 0)
	{			//when you find the key
	  count++;
	  return count;		//return the count
	}
      count++;
      N = N->next;		//if not move onto the next item on the list
    }
  return 0;			//if nothing is found return 0
}

//insert()
//parameters: Dictionary D, k, v
//post: none
//inserts new(key, value) into D
void
insert (Dictionary D, char *k, char *v)
{
  if (D == NULL)
    {				//if the dictionary pointer is null return a NULL dictionary reference error
      fprintf (stderr,
	       "Dictionary Error: insert() called on NULL Dictionary reference\n");
      exit (EXIT_FAILURE);
    }
  if ((lookup (D, k)) != NULL)
    {				//check to make sure we dont already have the key
      fprintf (stderr, "Dictionary Error: insert called on duplicate key\n");
      exit (EXIT_FAILURE);	//if we do produce the error and exit
    }
  if (D->numItems == 0)
    {				//if were at the start of the dictionary
      Node N = newNode (k, v);	//create a new node at the front
      D->head = N;		//then add it at the front of the dictionary
    }
  else
    {
      Node P = find (D, D->numItems);	//find the end of the dictionary
      P->next = newNode (k, v);	//add a new node right after it
      P = P->next;		//then move the node to that location so the new node is the end of the dictionary
    }
  D->numItems++;
}

//delete()
//parameters: Dictionary D, k
void
delete (Dictionary D, char *k)
{
  if (D == NULL)
    {				//check for a null dictionary reference
      fprintf (stderr,
	       "Dictionary Error: delete() called on NULL Dictionary reference\n");
      exit (EXIT_FAILURE);
    }
  if (lookup (D, k) == NULL)
    {				//make sure we are deleting a key
      fprintf (stderr,
	       "Dictionary Error: delete() called on non-existant key:\n");
      exit (EXIT_FAILURE);
    }
  int count = findIndex (D, k);	//find the index needed
 // if ((D->numItems == 1) && (strcmp (D->head->key, k) == 0))
   // {				//if the head is the only item in the list
     // free (D->head);		//just free the head
     // D->head = NULL;		//and point the pointer to NULL
   // }
 // else
 //   {				//otherwise the list is larger than 1
      if (strcmp (D->head->key, k) == 0)	//check if were at the start of the list
	{
	  Node P = D->head;	//set Node P to the head
	  D->head = D->head->next;	//move the head to the next node
	  freeNode (&P);	//free the old head
	}
      else
	{
	  Node P = find (D, count - 1);	//find the Node right before the one we need to delete
	  Node N = P->next;	//set the Node N to where the item we want to delete is
	  P->next = N->next;	//make the next item in the Node P to the next item in Node N, therefore getting rid of the item we need to delete
	  freeNode (&N);	//free the target Node
	}
   // }
  D->numItems--;		//reduce the number of items
}

//makeEmpty()
////returns Dictionary to empty state
////parameters: Dictionary D
////post: none
void
makeEmpty (Dictionary D)
{
  if (D == NULL)
    {				//check for a null dictionary reference
      fprintf (stderr,
	       "Dictionary Error: makeEmpty() called on NULL Dictionary reference\n");
      exit (EXIT_FAILURE);
    }
 /* if (D->numItems == 0)
    {				//if we already have an empty dictionary then produce an empty dictionary error
      fprintf (stderr,
	       "Dictionary Error: makeEmpty() called on empty Dictionary\n");
      D->head = NULL;
      free (D);
      exit (EXIT_FAILURE);
    }*/
  while (D->numItems > 0)
    {				//delete nodes until the number of items is 0
      delete (D, D->head->key);
    }
  D->head = NULL;		//set the head to null
  D->numItems = 0;		//set the number of items to 0
}



// printStack()
// prints a text representation of S to the file pointed to by out
// pre: none
void
printDictionary (FILE * out, Dictionary S)
{
  Node N;			//create a new node
  if (S == NULL)
    {				//check for a null dictionary reference error
      fprintf (stderr,
	       "Dictionary Error: printDictionary() called on NULL Dictionary reference\n");

      exit (EXIT_FAILURE);
    }
  for (N = S->head; N != NULL; N = N->next)
    fprintf (out, "%s %s\n", N->key, N->value);	//if not go from the head of the dictionary to the end of the dictionary printing the key and value with a newline
}

File: cs12b/pa3/backup/Dictionary.java
//Shayan Salsabilian
//ssalsabi
//2/13/19
//12B
//Creates a linked list called dictionary and creates basic functions for that linked list
//Dictionary.java
public class Dictionary implements DictionaryInterface
{
	class Node
	{
		String key;			//created two string classes to hold the key value
		String value;
		Node next;			//then create a next node to continue the list

		Node (String x, String y)	//what needs to be placed when creating a new node
		{
			key = x;			//the key will be the x when creating a new node
			value = y;		//the value will be the y when creating a new node
			next = null;		//and the next item will be null till we connect another node
		}
	}

	// Fields for the IntegerList class
	private Node head;		// reference to first Node in List
	private int numItems;		// number of items in this IntegerList

	//IntegerList()
	// constructor for the IntegerList class
	public Dictionary ()
	{
		head = null;		//the class dictionary starts with a empty node with 0 items
		numItems = 0;
	}


	// find() (helper function)
	// returns a reference to the Node at position index in the Dictionary
	private Node find (int index)
	{
		Node N = head;
		for (int i = 1; i < index; i++)
		{				//keep going till we reach the index were looking for
			N = N.next;
		}
		return N;
	}

	// isEmpty() (ADT function)
	// pre: none
	// post: returns true if this Dictionary is empty, false otherwise
	public boolean isEmpty ()
	{
		return (numItems == 0);	//conditional with when numItems is 0
	}

	// size() (ADT function)
	// pre: none
	// post: returns the number of elements in this Dictionary
	public int size ()
	{
		return numItems;		//return the numItems that is the size
	}

	// insert()  (ADT function)
	// pre: key
	// post: returns value at the key in this Dictionary
	public String lookup (String key)
	{
		Node N = head;		//have a Node point at the front of the list
		for (int i = 1; i <= numItems; i++)
		{				//go through each item in the Dictionary till we reach the max number of items
			if ((N.key.compareTo (key)) == 0)	//compare the key to the key in each of those parts of the Dictionary
			{			//when one matches up that means were at the proper key in the Dictionary
				return N.value;	//then we can return the value because were at the proper key
			}
			N = N.next;		// if the key doesnt match up to that part of the Dictionary, go to the next part of the dictionary
		}
		return null;		//if we have gone through the entire list and havent found it return null
	}

	// findIndex (helper function)
	// finds the position of the key within the Dictionary
	// pre: key
	// post: returns an int containing how far away we are from the front of the list
	private int findIndex (String key)
	{
		int count = 0;		//start the count at zero
		Node N = head;		//start the Node at the front of the list
		while (count <= numItems)
		{				//go through the entire Dictionary
			if ((N.key.compareTo (key)) == 0)	//once we find a part of the Dictionary that matches the key 
			{
				count++;		//increment the count (since lists start at 1)
				return count;	//then return the count
			}
			count++;		//otherwise increment the count and go to the next key in the Dictionary
			N = N.next;
		}
		return 0;			//if we dont find the key in the list just return a 0, cant be null because int return
	}
	// insert () (ADT function)
	// inserts new key and value into the dictionary at the end
	// pre: key, value
	// post: none
	public void insert (String key, String value) throws DuplicateKeyException
	{
		if ((lookup (key)) != null)	//looks through the list and makes sure the key doesnt already exist, if it does throw a duplicate key exception
		{
			throw new DuplicateKeyException ("cannot insert duplicate keys");
		}
		if (numItems == 0)		//if we are inserting the very first item in the list
		{
			Node N = new Node (key, value);	// create a new node with a key and value
			head = N;		//make that new node be the head of the Dictionary
		}
		else
		{
			Node P = find(numItems); //find were we are in the dictionary and return that node
			Node N=new Node(key, value);
			//P.next = new Node (key, value);	//make that next node a new node with the key and value needed
			N.next= P.next;
			P.next= N;
			N = N.next;		//then move to the point in the dictionary making it the new end of the list
		}
		numItems++;			//increment the number of items
	}

	// delete() (ADT function)
	// delete node from there position in the Dictionary using the key to determine where that is
	// pre: key
	// post: none
	Node findKey(String key){//didn't use but in case someone else did
		Node R=head;
		while(R.next!=null){
			if((R.next.key.compareTo(key))==0){
				return R; 
			}
			R=R.next; 
		}
		return R;
	}

	public void delete (String key) throws KeyNotFoundException
	{
		if ((lookup (key)) == null)	//check to make sure the key exists
		{
			throw new KeyNotFoundException ("cannot delete non-existent key");	//if it doesnt print an error
		}
		int count = findIndex (key);	//find the position of the key in the dictionary
		if (count==1)		//if we are in position 1 in the dictionary
		{
			head = head.next;	//just move the head to position 2 in the dictionary losing the pointer to the start of the dictionary
		}
		else
		{
			Node P = find(count-1);	//find the node at the position right before the key
			P.next = P.next.next;	//change the node that contains the key, so that it points to the next item after the key (ie N.next since N points at the key)
			
		}
		numItems--;			//decrement the number of items
	}

	// makeEmpty() (ADT function)
	// pre: none
	// post: isEmpty()
	public void makeEmpty ()
	{
		head = null;		//lose the pointer that contains heads
		numItems = 0;		//make the number of items equal to 0
	}

	// toString()
	// pre: none
	// post:  prints current state to stdout
	// Overrides Object's toString() method
	public String toString ()
	{
		StringBuffer sb = new StringBuffer ();
		Node N = head;		//start from the head 
		for (; N != null; N = N.next)
		{				//go through the dictionary
			sb.append (N.key).append (" ").append (N.value).append ("\n");	//make each line have the key with a space then the value followed by the newline
		}
		return new String (sb);	//return the string
	}

}
